File: web/src/lib/disks.ts
// Pure disk-entity display-shaping helpers for the Storage view: which
// order disk cards render in, what "role" group a slot's name implies,
// and how to read its temp/spin-state -- kept in their own module
// (matching containersSort.ts/topFromFrame.ts's one-file-per-view-concern
// convention) rather than folded into metrics.ts's more general,
// disk-agnostic helpers.
import type { SeriesPoint } from './api';

export type DiskRole = 'parity' | 'data' | 'pool' | 'flash';

// diskRole classifies a disk ENTITY NAME into one of Storage's four grid
// groups. There's no dedicated "is this a pool" field anywhere this
// collector reads (see ArrayCard's own doc on the same gap) -- a slot is
// "pool" by ELIMINATION: not parity, not a numbered disk*, not the
// literal boot "flash" device. That correctly buckets both the literal
// "cache" slot and a real named pool (e.g. "rocket_pool", per
// fixtures.md) the same way, with no hardcoded name list to keep in
// sync with a real array's actual pool names.
export function diskRole(name: string): DiskRole {
  if (name.startsWith('parity')) return 'parity'; // "parity", "parity2" (dual parity)
  if (/^disk\d+$/.test(name)) return 'data';
  if (name === 'flash') return 'flash';
  return 'pool';
}

const ROLE_RANK: Record<DiskRole, number> = { parity: 0, data: 1, pool: 2, flash: 3 };

// naturalSuffix extracts a name's trailing digit run, for a numeric-
// aware sort ("disk2" before "disk10", "parity" before "parity2") --
// null for a name with no trailing digits (every pool name, "flash"),
// which falls back to plain alpha ordering instead.
function naturalSuffix(name: string): number | null {
  const m = /(\d+)$/.exec(name);
  return m ? parseInt(m[1], 10) : null;
}

// sortDiskEntities orders disk entity names into the Storage grid's
// fixed group order (parity* first, then disk* by number, then cache/
// pools, then the boot flash device last), breaking every tie by name
// ascending for deterministic, reproducible output.
export function sortDiskEntities(names: string[]): string[] {
  return [...names].sort((a, b) => {
    const ra = ROLE_RANK[diskRole(a)];
    const rb = ROLE_RANK[diskRole(b)];
    if (ra !== rb) return ra - rb;
    const na = naturalSuffix(a);
    const nb = naturalSuffix(b);
    if (na !== null && nb !== null && na !== nb) return na - nb;
    return a.localeCompare(b);
  });
}

export type DiskMediaType = 'hdd' | 'ssd';

// diskMediaType reads a disk's rotational value (disks.go's own "0/1 per
// present disk from disks.ini" contract) into the two-way distinction
// Storage/Overview render a glyph for: 0 is solid-state, anything else
// (1, the only other real-world value, but any nonzero reading is
// treated the same) is spinning. null -- absent, never guessed -- covers
// a not-present disk (never rendered here anyway) and a stale/pre-
// upgrade frame that hasn't started sending this metric yet.
//
// Superseded by diskKind below as Storage/Overview's own primary read
// (rotational alone can't tell a USB flash stick or an NVMe pool member
// apart from an ordinary spinning/SATA-SSD disk -- Scott's own report),
// but kept, tested, and exported as-is: diskKind falls back to it
// whenever a frame's disk_meta doesn't (yet) cover a slot.
export function diskMediaType(metrics: Record<string, number> | undefined | null): DiskMediaType | null {
  const rotational = metrics?.['rotational'];
  if (rotational === undefined) return null;
  return rotational === 0 ? 'ssd' : 'hdd';
}

export type DiskKind = 'hdd' | 'ssd' | 'nvme' | 'usb';

const DISK_KINDS: ReadonlySet<string> = new Set<DiskKind>(['hdd', 'ssd', 'nvme', 'usb']);

// diskMetaKind narrows an arbitrary string (straight off the wire, never
// typechecked at its source) to DiskKind -- same "don't trust the network"
// convention topFromFrame.ts's isTopResource already uses for a route
// param. Exported: containerStorage's own device rows reuse this exact
// narrowing for DeviceIODTO's Kind (unraid.ResolveDeviceLabel's own
// hdd/ssd/nvme/usb vocabulary, straight off the same disk_meta-derived
// source) rather than re-declaring the DISK_KINDS set a second time.
export function diskMetaKind(value: string | undefined): DiskKind | null {
  return value !== undefined && DISK_KINDS.has(value) ? (value as DiskKind) : null;
}

// diskKind is Storage/Overview's own primary type read: the server now
// classifies every present disk (unraid.DiskKind, disks.go) into one of
// four kinds -- rotational alone can conflate a USB flash stick with an
// ordinary spinning disk, and an NVMe pool member with a plain SATA SSD,
// which is exactly the bug report this exists to fix. meta is that
// slot's own disk_meta entry (SnapshotDTO.DiskMeta, keyed by slot);
// falling back to the older rotational-only diskMediaType (never hidden
// outright) covers a frame from before this feature shipped, or a slot
// disk_meta hasn't (yet) reported on this tick.
export function diskKind(
  meta: { kind?: string } | undefined | null,
  metrics: Record<string, number> | undefined | null,
): DiskKind | null {
  return diskMetaKind(meta?.kind) ?? diskMediaType(metrics);
}

export type DiskTempState = { kind: 'reading'; celsius: number } | { kind: 'spun-down' } | { kind: 'no-sensor' };

// diskTempState reads one disk's temp/spin-state metrics. Absence of the
// "temp.c" key is itself the signal a real spun-down disk gives (Unraid
// never reports a temp for a sleeping drive -- disks.go's own doc) --
// this checks KEY PRESENCE, not mere falsiness, since 0C is a
// theoretically valid reading that must never be confused with "no
// sample at all". spun_up===0 explains the absence as an expected sleep
// state ("spun-down"); anything else -- spun_up===1, or spun_up itself
// absent -- reads as "no-sensor" instead: the real boot flash device is
// the documented case (reports spundown="0" but temp="*", i.e. no
// sensor at all, unrelated to spin state -- see fixtures.md discrepancy
// 11), and a disk with no spundown key at all degrades the same safe
// way rather than guessing.
export function diskTempState(metrics: Record<string, number> | undefined | null): DiskTempState {
  const temp = metrics?.['temp.c'];
  if (temp !== undefined) return { kind: 'reading', celsius: temp };
  return metrics?.['spun_up'] === 0 ? { kind: 'spun-down' } : { kind: 'no-sensor' };
}

// diskUsagePct computes a disk's fill percentage from its fs bytes pair,
// or null when either half is absent (e.g. parity, which has no
// filesystem view at all -- disks.go's own doc).
export function diskUsagePct(metrics: Record<string, number> | undefined | null): number | null {
  const used = metrics?.['fs.used_bytes'];
  const free = metrics?.['fs.free_bytes'];
  if (used === undefined || free === undefined) return null;
  const total = used + free;
  return total > 0 ? (used / total) * 100 : 0;
}

// diskUsagePctSeries is diskUsagePct's history-shaped sibling, for the
// Storage chart's "Used" line: fs.used_bytes and fs.free_bytes are two
// separate /api/series results (there's no server-side usage-percent
// series), zipped point-by-point by exact ts -- both are recorded on
// the same collector tick, so a ts missing from one side (a store gap)
// is skipped rather than guessed at.
export function diskUsagePctSeries(usedPoints: SeriesPoint[], freePoints: SeriesPoint[]): [number, number][] {
  const freeByTs = new Map(freePoints.map(([ts, avg]) => [ts, avg]));
  const out: [number, number][] = [];
  for (const [ts, used] of usedPoints) {
    const free = freeByTs.get(ts);
    if (free === undefined) continue;
    const total = used + free;
    out.push([ts, total > 0 ? (used / total) * 100 : 0]);
  }
  return out;
}

// defaultDiskChartVisible decides the storage chart's own starting
// legend state ("keep 12+ lines calm"): pools and parity default
// visible regardless of activity (the array's own backbone -- and
// parity carries no usage/IO of its own to judge activity by anyway),
// an ordinary data/flash disk only if it's actually doing something
// right now; everything else starts toggled off.
export function defaultDiskChartVisible(slot: string, hasRecentIO: boolean): boolean {
  const role = diskRole(slot);
  return role === 'pool' || role === 'parity' || hasRecentIO;
}

File: internal/collect/unraid/disks.go
package unraid

import (
	"fmt"
	"log"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"

	"github.com/smidley/gantry/internal/store"
)

// diskNotPresent is the status-value prefix disks.ini uses for an empty
// array slot (observed real forms: "DISK_NP" and "DISK_NP_DSBL"). Every
// other status (DISK_OK today; disabled/invalid states in later Unraid
// versions) counts as present — per-disk metrics gate on "not DISK_NP",
// not on an allowlist of known-good statuses.
const diskNotPresent = "DISK_NP"

// DiskMeta is one present disk slot's device name and classified type
// (see DiskKind) — strings, so unlike rotational/temp/etc. they can't
// ride the numeric MetricSink. Collector.DiskMeta() exposes the latest
// tick's own map for the snapshot builder to merge into
// server.SnapshotDTO.DiskMeta, mirroring Version()'s identical "set on
// tick under mu, read via a copying getter" shape, for the same
// cross-goroutine reason (the snapshot builder runs on a different
// goroutine than Tick).
type DiskMeta struct {
	Device string
	Kind   string
}

// DiskKind classifies a present disk slot into the frontend's four-way
// type badge: "usb" for the boot flash device, "nvme" for an NVMe pool
// member, "ssd" for any other solid-state member, "hdd" for a spinning
// one. Exported and pure (no *Collector receiver) so both a real box's
// own tickOneDisk below and the fake generator's disk metadata share one
// tested rule.
//
// Neither signal alone is enough (Scott's own report, reproduced against
// testdata/disks_real.ini): the boot flash device reports rotational=1
// AND a plain SCSI-style device name ("sdi") — indistinguishable from a
// real spinning disk by either signal on its own — and an NVMe pool
// member reports rotational=0, the same as a plain SATA/SAS SSD; only
// the device name ("nvme0n1" vs "sdX") tells THOSE two apart. Order
// matters: the boot device's SLOT NAME ("flash", Unraid's own fixed,
// version-independent name for it) is checked first and wins outright,
// before either signal below ever looks at this slot's own device name
// or rotational value. An unparseable/absent rotational reading (should
// not happen for a present disk in practice) falls back to "hdd", the
// least alarming default.
func DiskKind(slot, device string, rotational float64, rotationalOK bool) string {
	if slot == "flash" {
		return "usb"
	}
	if strings.HasPrefix(device, "nvme") {
		return "nvme"
	}
	if rotationalOK && rotational == 0 {
		return "ssd"
	}
	return "hdd"
}

// tickDisks reads disks.ini and records per-present-disk stats, emitting
// a disk.errors event whenever a slot's error count has risen since the
// last tick it was seen. Missing/unreadable disks.ini degrades silently
// (no error) — it isn't Probe's hard dependency, only var.ini is.
func (c *Collector) tickDisks(now time.Time) {
	f, err := os.Open(filepath.Join(c.dir, "disks.ini"))
	if err != nil {
		return
	}
	defer func() { _ = f.Close() }()

	kv, err := ParseINI(f)
	if err != nil {
		return
	}

	slots := make([]string, 0, len(kv))
	for slot := range kv {
		if slot == "" {
			continue // disks.ini has no headerless keys; stay tolerant if it ever did
		}
		slots = append(slots, slot)
	}
	sort.Strings(slots)

	c.recordSlots(slots, kv)

	ts := now.Unix()
	for _, slot := range slots {
		c.tickOneDisk(slot, kv[slot], ts)
	}
}

// recordSlots classifies every slot by its "type" field ("Cache" -> pool;
// anything else -- "Data", "Parity", "Flash", absent, or unrecognized --
// is not a pool) and stores the list for Slots(). slots is already
// sorted, so the output list stays sorted too. Runs regardless of a
// slot's present/DISK_NP status: an empty bay is still a known part of
// the fleet, unlike tickOneDisk's metrics (which have nothing
// meaningful to report for a slot with no filesystem).
func (c *Collector) recordSlots(slots []string, kv map[string]map[string]string) {
	var pools []string
	for _, slot := range slots {
		if kv[slot]["type"] == "Cache" {
			pools = append(pools, slot)
		}
	}
	c.mu.Lock()
	c.poolSlots = pools
	c.mu.Unlock()
}

func (c *Collector) tickOneDisk(slot string, disk map[string]string, ts int64) {
	if strings.HasPrefix(disk["status"], diskNotPresent) {
		return
	}

	if spundown, ok := parseFloatOK(disk["spundown"]); ok {
		c.sink.Record(store.SeriesKey{Kind: "disk", Entity: slot, Metric: "spun_up"}, ts, 1-spundown)
	}

	// rotational/rotationalOK feed both the numeric metric below (recorded
	// only when parseable, same as every other metric here) and DiskKind's
	// classification just after (which needs to know when it's missing,
	// not merely treat an unparseable reading as some default value).
	rotational, rotationalOK := parseFloatOK(disk["rotational"])
	if rotationalOK {
		c.sink.Record(store.SeriesKey{Kind: "disk", Entity: slot, Metric: "rotational"}, ts, rotational)
	}

	device := disk["device"]
	c.mu.Lock()
	c.diskMeta[slot] = DiskMeta{Device: device, Kind: DiskKind(slot, device, rotational, rotationalOK)}
	c.mu.Unlock()

	// temp is a number, or the literal "*" when the disk is spun down or
	// its temp is otherwise unknown; parseFloatOK's failure on "*" is
	// exactly the signal we want — omit the sample entirely rather than
	// ever recording a misleading 0.
	if temp, ok := parseFloatOK(disk["temp"]); ok {
		c.sink.Record(store.SeriesKey{Kind: "disk", Entity: slot, Metric: "temp.c"}, ts, temp)
	}

	if numErrors, ok := parseFloatOK(disk["numErrors"]); ok {
		c.sink.Record(store.SeriesKey{Kind: "disk", Entity: slot, Metric: "errors"}, ts, numErrors)
		c.checkErrorIncrease(slot, numErrors)
	}

	// fsSize/fsFree are KB-unit strings and are "0" for a disk with no
	// filesystem view (parity); only emit the byte metrics when there's
	// an actual filesystem size to report, and only when both parsed.
	// fsUsed is authoritative when present (real btrfs pools were observed
	// to diverge from fsSize-fsFree); fall back to the subtraction for a
	// disks.ini that doesn't carry it.
	fsSize, sizeOK := parseFloatOK(disk["fsSize"])
	fsFree, freeOK := parseFloatOK(disk["fsFree"])
	if sizeOK && freeOK && fsSize > 0 {
		fsUsed, usedOK := parseFloatOK(disk["fsUsed"])
		if !usedOK {
			fsUsed = fsSize - fsFree
		}
		c.sink.Record(store.SeriesKey{Kind: "disk", Entity: slot, Metric: "fs.used_bytes"}, ts, fsUsed*1024)
		c.sink.Record(store.SeriesKey{Kind: "disk", Entity: slot, Metric: "fs.free_bytes"}, ts, fsFree*1024)

		// fs.used_pct: byte-identical to web/src/lib/disks.ts:diskUsagePct
		// (used/(used+free)*100) -- fsUsed/fsFree are still in their raw
		// KB unit here, but the ratio is scale-invariant, so there's no
		// need to redo this against the *1024 values just recorded above.
		// Persisted (no "live:" prefix) so the alert engine's
		// disk-usage-high rule and the Storage view's display band read
		// the exact same number the frontend used to derive only for
		// itself, client-side.
		if total := fsUsed + fsFree; total > 0 {
			c.sink.Record(store.SeriesKey{Kind: "disk", Entity: slot, Metric: "fs.used_pct"}, ts, fsUsed/total*100)
		}
	}
}

// checkErrorIncrease emits a disk.errors alert event the first time (per
// slot) numErrors is seen to have risen since the previous tick that slot
// was observed. A slot seen for the first time only seeds
// prevDiskErrors silently: its possibly-already-nonzero count reflects
// history from before this collector started, not something that just
// happened.
func (c *Collector) checkErrorIncrease(slot string, numErrors float64) {
	prev, seen := c.prevDiskErrors[slot]
	c.prevDiskErrors[slot] = numErrors
	if !seen || numErrors <= prev {
		return
	}
	_, err := c.events.AppendEvent(store.Event{
		Kind: "disk.errors", Entity: slot, Severity: "alert",
		Detail: fmt.Sprintf("errors %g → %g", prev, numErrors),
	})
	if err != nil {
		log.Printf("events: %v", err)
	}
}
